### ISSUE TITLE
Request- Deprecation warnings and stability between releases

### ISSUE THREAD

**Issue by ClimbsRocks** [Microsoft/LightGBM#967]
Hi Team,

This package is awesome, and we use it a lot. It's become our default gradient boosting implementation at work. The training speed in particular is astounding. 

I was disappointed to find that the latest release had two changes that broke our existing code: The type of data inputs accepted changed, throwing `ValueError`s, and the `.best_iteration` attribute disappeared, throwing an `AttributeError`. Unfortunately, I remember something pretty similar from the previous patch release, though I forget exactly what the issue was now. 

I'm all for progress! One of the things I like about this package is how rapidly everyone's developing on it. But it's marked as `'Development Status :: 5 - Production/Stable'`, and that means that code that's written using lightgbm one patch ago (`2.0.6`) should definitely work with this patch (`2.0.7`). 

Generally, I'd expect Deprecation warnings, or FutureWarnings, or UserWarnings for at least 2 minor versions (`2.0.x - > 2.2.x`), and ideally one major version, before a release that breaks existing codebases. 

I know that it's a bit more work. It possibly means setting up a test environment that runs your tests on multiple releases of your code. It definitely means a couple more if statements to check for things like if the user is trying to access a soon-to-be-removed attribute, and returning to them the new attribute, and a warning. 

But that's what you signed up for with `'Development Status :: 5 - Production/Stable'`. If you want to be able to release breaking changes, there are other Development Statuses for that, and we'll adjust our code to be more cautious accordingly. 

Again, thanks for the great package! It's vastly sped up our iteration speed at work; it's as accurate as anything else out there that we've found; it's got great use in analytics projects too with `.feature_importances_`. I'm a loyal evangelist.

**Comment by wxchan** [comment]
Thanks for your suggestions. We don't have much experience on this and will be careful in the future. 

Btw, what's exactly the two changes you refer to (`The type of data inputs accepted changed` & `the .best_iteration attribute disappeared`)? Do you mean in sklearn interface?

**Comment by StrikerRUS** [comment]
Sorry that you had problems with LightGBM's backward compatibility. From my side I confirm that
>we will do our best in the future.

As a typical user, I can remind you that you can fix versions of dependences of your software. It's very useful when you have dozens of them. And then you periodically do review of dependences and in case of "all OK" increment the version of the corresponding dependence.

**Comment by ClimbsRocks** [comment]
Thanks for the responses, y'all! 

@wxchan you nailed it- i'm using the sklearn interface. 

as always, with open source software, it's all a bit buyer beware, but i thought i'd mention it so y'all knew how much people were using and depending on your project. 

thanks always for the great work. we're very glad to be using it :)
